src/client/particles.cpp

```diff
@@ -38,105 +38,107 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 Particle::Particle(
-	IGameDef *gamedef,
-	LocalPlayer *player,
-	ClientEnvironment *env,
-	const ParticleParameters &p,
-	const ClientTexRef& texture,
-	v2f texpos,
-	v2f texsize,
-	video::SColor color
-):
-	scene::ISceneNode(((Client *)gamedef)->getSceneManager()->getRootSceneNode(),
-		((Client *)gamedef)->getSceneManager()),
-	m_texture(texture)
+		IGameDef *gamedef,
+		LocalPlayer *player,
+		ClientEnvironment *env,
+		const ParticleParameters &p,
+		const ClientParticleTexRef &texture,
+		v2f texpos,
+		v2f texsize,
+		video::SColor color,
+		ParticleSpawner *parent,
+		std::unique_ptr<ClientParticleTexture> owned_texture
+	) :
+		scene::ISceneNode(((Client *)gamedef)->getSceneManager()->getRootSceneNode(),
+				((Client *)gamedef)->getSceneManager()),
+
+		m_expiration(p.expirationtime),
+
+		m_env(env),
+		m_gamedef(gamedef),
+		m_collisionbox(aabb3f(v3f(-p.size / 2.0f), v3f(p.size / 2.0f))),
+		m_texture(texture),
+		m_texpos(texpos),
+		m_texsize(texsize),
+		m_pos(p.pos),
```

Comment:
Idea: make `Particle` inherit or contain `ParticleParameters` which would then allow you to easily copy `p` into `this`.

Reply:
I had the same idea, but I'm not sure if it's good to do this.

I've made a commit to try it out. I'll revert or not based on your opinion.

There are some unused members:
* `texture` (a string)
* `node` (a MapNode)
* `node_tile` (u8)

Also, I've made the params field const (because it would feal weird to modify the params after creation, and also someone might want to optimize in the future by removing the copy and instead add a reference to the params of the spawner).
Because of this there are these fields that are duplicated (non of them are in `CommonParticleParams`):
* `pos, vel, acc` (acc might also be const, but I preferred consistency)
* `expirationtime` (The only write to it is for a somewhat hacky forced expiration at collision.)